src/typechecker.erl

```diff
@@ -4426,6 +4438,19 @@ is_list_pat_exhaustive({nil, _}) -> true;
 is_list_pat_exhaustive({cons, _, {var, _, _}, {var, _, _}}) -> true;
 is_list_pat_exhaustive(_) -> false.
 
+%% TODO: This is incomplete!
+%% To properly check pattern exhaustiveness we have to consider bound variables.
```

Comment:
A nice way to remember this is to add a known problem where we e.g. have a bound variable V in `<<..., V/binary>>`.

Reply:
The latter actually leads to:

```
t2.erl: The variable on line 13 at column 11 is expected to have type non_neg_integer() but it has type <<_:_*8>>
```

but indeed, they don't lead to any exhaustiveness false positives, which is a good thing.

> since the real logic is in type_diff?

For the case when `should_bin_pat_enable_exhaustiveness_check()` returns true and the examples I could think of or you provide - yes. If `should_bin_pat_enable_exhaustiveness_check()` returns false we disable exhaustiveness checking for the clauses list completely, therefore avoid any false positives.

---

Ultimately, because of https://github.com/josefs/Gradualizer/pull/392#issuecomment-1145871376 we could just disable exhaustiveness checking for binary patterns completely (i.e. keep `refinable(binary()) == true`, but disable ex-check each time we see a binary pattern), and the goal to be able to include them in complex types would be met anyway. That would be as simple as possible, but the tradeoff would be that even the simplest binary patterns (for example an empty binary) would disable the ex-check on the complex type, too.

Reply:
Thanks for the explanation.

Btw, I think my g/2 example could easily be corrected, e.g. `<<Y:1/binary, _/binary>>`.

Reply:
If you leave out the pattern `<<>>` in the examples `f/2` and the corrected `g/2`, you'd get "Values not covered: `<<>>`" and when you add a clause for `<<>>`, the error dissapears, which can make the user believe all values are covered. Hopefully, they have RTFM and found that exhaustiveness-checking is best effort (i.e. unsound) anyway, so all is fine. :-)